### ISSUE TITLE
API server overloaded, apparently due to dns autoscaler config changes

### ISSUE THREAD

**Issue by porridge** [kubernetes/kubernetes#55779]
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug
/priority failing-test

**What happened**:
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-correctness/28

```
[sig-cli] Kubectl client [k8s.io] Update Demo should create and stop a replication controller [Conformance] 1m37s
go run hack/e2e.go -v -test --test_args='--ginkgo.focus=\[sig\-cli\]\sKubectl\sclient\s\[k8s\.io\]\sUpdate\sDemo\sshould\screate\sand\sstop\sa\sreplication\scontroller\s\s\[Conformance\]$'
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:647
Expected error:
    <exec.CodeExitError>: {
        Err: {
            s: "error running &{/workspace/kubernetes/platforms/linux/amd64/kubectl [kubectl --server=https://104.196.181.1 --kubeconfig=/workspace/.kube/config get pods update-demo-nautilus-874l7 -o template --template={{if (exists . \"status\" \"containerStatuses\")}}{{range .status.containerStatuses}}{{if (and (eq .name \"update-demo\") (exists . \"state\" \"running\"))}}true{{end}}{{end}}{{end}} --namespace=e2e-tests-kubectl-mzjqk] []  <nil>  Unable to connect to the server: net/http: TLS handshake timeout\n [] <nil> 0xc43e12e960 exit status 1 <nil> <nil> true [0xc42d1f8038 0xc42d1f8050 0xc42d1f8070] [0xc42d1f8038 0xc42d1f8050 0xc42d1f8070] [0xc42d1f8048 0xc42d1f8068] [0x999d80 0x999d80] 0xc420a3e180 <nil>}:\nCommand stdout:\n\nstderr:\nUnable to connect to the server: net/http: TLS handshake timeout\n\nerror:\nexit status 1\n",
        },
        Code: 1,
    }
    error running &{/workspace/kubernetes/platforms/linux/amd64/kubectl [kubectl --server=https://104.196.181.1 --kubeconfig=/workspace/.kube/config get pods update-demo-nautilus-874l7 -o template --template={{if (exists . "status" "containerStatuses")}}{{range .status.containerStatuses}}{{if (and (eq .name "update-demo") (exists . "state" "running"))}}true{{end}}{{end}}{{end}} --namespace=e2e-tests-kubectl-mzjqk] []  <nil>  Unable to connect to the server: net/http: TLS handshake timeout
     [] <nil> 0xc43e12e960 exit status 1 <nil> <nil> true [0xc42d1f8038 0xc42d1f8050 0xc42d1f8070] [0xc42d1f8038 0xc42d1f8050 0xc42d1f8070] [0xc42d1f8048 0xc42d1f8068] [0x999d80 0x999d80] 0xc420a3e180 <nil>}:
    Command stdout:
    
    stderr:
    Unable to connect to the server: net/http: TLS handshake timeout
    
    error:
    exit status 1
    
not to have occurred
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:2094
				
```
@kubernetes/sig-cli-bugs please investigate ASAP. This is a release-blocking test group.

**Comment by porridge** [comment]
Umbrella issue - #55194

**Comment by porridge** [comment]
Dear sig-cli leads @fabianofranz, @pwittrock, @AdoHe: ping?

**Comment by porridge** [comment]
/priority critical-urgent
this blocks #55194

**Comment by dixudx** [comment]
> Unable to connect to the server: net/http: TLS handshake timeout

@porridge kubectl cannot connect to apiserver. This may be not a bug of kubectl.

**Comment by porridge** [comment]
That's possible. Can you figure out at what timestamp this happened and what the api server was doing at that time? Could it be that the test or kubectl itself needs some retries?

**Comment by dixudx** [comment]
> Could it be that the test or kubectl itself needs some retries?

kubectl already retried after 2s before dying.

https://github.com/kubernetes/kubernetes/blob/44f24d219f9d28121eb1f16863312f790e144882/test/e2e/framework/util.go#L2084-L2098

The `net/http: TLS handshake timeout` indicated the connection was bad. Probably apiserver was not working.

**Comment by porridge** [comment]
On Tue, Nov 21, 2017 at 9:52 AM, Di Xu <notifications@github.com> wrote:

> The net/http: TLS handshake timeout indicated the connection was bad.
> Probably apiserver was not working.
>
>
To debug this, we need a timestamp of when the timeout happened. Can you
find it in the log?

**Comment by mengqiy** [comment]
@porridge It seems it stops failing.

**Comment by porridge** [comment]
@mengqiy yes, this test fails intermittently, but the problem is that *some* test is failing in this job pretty much every single time, and it is not a good state to be in for a release-blocking job.
So just saying "oh, it was just this one-time flake" is not an option. We're in this "surely it will pass the next time" mode for over 2 weeks now.
We need to make progress on fixing the flakes, even if the progress is "we'll add this debug statements here to make this possible to root-cause the next time it happens".

And due to the sheer number of problems (see the number of references on #55194), we really need to delegate some work to sigs responsible for the tests, we will not be able to handle all of them ourselves in sig-scalability. Please help us out here.

**Comment by porridge** [comment]
This is the relevant piece of the test output:

```
I1114 21:39:54.551] Nov 14 21:38:41.093: INFO: Running '/workspace/kubernetes/platforms/linux/amd64/kubectl --server=https://104.196.181.1 --kubeconfig=/workspace/.kube/config get pods -o template --template={{range.items}}{{.metadata.name}} {{end}} -l name=update-demo --namespace=e2e-tests-kubectl-mzjqk'
I1114 21:39:54.552] Nov 14 21:38:41.424: INFO: stderr: ""
I1114 21:39:54.552] Nov 14 21:38:41.424: INFO: stdout: "update-demo-nautilus-874l7 update-demo-nautilus-vb9gp "
I1114 21:39:54.552] Nov 14 21:38:41.424: INFO: Running '/workspace/kubernetes/platforms/linux/amd64/kubectl --server=https://104.196.181.1 --kubeconfig=/workspace/.kube/config get pods update-demo-nautilus-874l7 -o template --template={{if (exists . "status" "containerStatuses")}}{{range .status.containerStatuses}}{{if (and (eq .name "update-demo") (exists . "state" "running"))}}true{{end}}{{end}}{{end}} --namespace=e2e-tests-kubectl-mzjqk'
I1114 21:39:54.552] Nov 14 21:38:41.758: INFO: stderr: ""
I1114 21:39:54.552] Nov 14 21:38:41.758: INFO: stdout: ""
I1114 21:39:54.553] Nov 14 21:38:41.758: INFO: update-demo-nautilus-874l7 is created but not running
I1114 21:39:54.553] Nov 14 21:38:46.758: INFO: Running '/workspace/kubernetes/platforms/linux/amd64/kubectl --server=https://104.196.181.1 --kubeconfig=/workspace/.kube/config get pods -o template --template={{range.items}}{{.metadata.name}} {{end}} -l name=update-demo --namespace=e2e-tests-kubectl-mzjqk'
I1114 21:39:54.553] Nov 14 21:38:48.651: INFO: stderr: ""
I1114 21:39:54.553] Nov 14 21:38:48.652: INFO: stdout: "update-demo-nautilus-874l7 update-demo-nautilus-vb9gp "
I1114 21:39:54.553] Nov 14 21:38:48.652: INFO: Running '/workspace/kubernetes/platforms/linux/amd64/kubectl --server=https://104.196.181.1 --kubeconfig=/workspace/.kube/config get pods update-demo-nautilus-874l7 -o template --template={{if (exists . "status" "containerStatuses")}}{{range .status.containerStatuses}}{{if (and (eq .name "update-demo") (exists . "state" "running"))}}true{{end}}{{end}}{{end}} --namespace=e2e-tests-kubectl-mzjqk'
I1114 21:39:54.554] Nov 14 21:39:11.441: INFO: rc: 127
I1114 21:39:54.554] Nov 14 21:39:11.441: INFO: stdout: ""
I1114 21:39:54.554] STEP: using delete to clean up resources
I1114 21:39:54.554] Nov 14 21:39:11.442: INFO: Running '/workspace/kubernetes/platforms/linux/amd64/kubectl --server=https://104.196.181.1 --kubeconfig=/workspace/.kube/config delete --grace-period=0 --force -f - --namespace=e2e-tests-kubectl-mzjqk'
I1114 21:39:54.554] Nov 14 21:39:28.254: INFO: stderr: "warning: Immediate deletion does not wait for confirmation that the running resource has been terminated. The resource may continue to run on the cluster indefinitely.\n"
I1114 21:39:54.555] Nov 14 21:39:28.254: INFO: stdout: "replicationcontroller \"update-demo-nautilus\" deleted\n"
I1114 21:39:54.555] Nov 14 21:39:28.254: INFO: Running '/workspace/kubernetes/platforms/linux/amd64/kubectl --server=https://104.196.181.1 --kubeconfig=/workspace/.kube/config get rc,svc -l name=update-demo --no-headers --namespace=e2e-tests-kubectl-mzjqk'
I1114 21:39:54.555] Nov 14 21:39:28.660: INFO: stderr: "No resources found.\n"
I1114 21:39:54.555] Nov 14 21:39:28.660: INFO: stdout: ""
I1114 21:39:54.555] Nov 14 21:39:28.660: INFO: Running '/workspace/kubernetes/platforms/linux/amd64/kubectl --server=https://104.196.181.1 --kubeconfig=/workspace/.kube/config get pods -l name=update-demo --namespace=e2e-tests-kubectl-mzjqk -o go-template={{ range .items }}{{ if not .metadata.deletionTimestamp }}{{ .metadata.name }}{{ "\n" }}{{ end }}{{ end }}'
I1114 21:39:54.555] Nov 14 21:39:29.772: INFO: stderr: ""
I1114 21:39:54.556] Nov 14 21:39:29.772: INFO: stdout: ""
```

**Comment by porridge** [comment]
Something bad clearly happened to the apiserver around the time when the error was returned - request response rate dropped from ~1k qps to nearly zero for a few good seconds:

```
porridge@kielonek:~/projects/k8s/55779$ egrep '^.1114 21:(38:[45]|39:[123])' kube-apiserver.log |cut -c 2-14|uniq -c
   1506 1114 21:38:40
   1816 1114 21:38:41
   1292 1114 21:38:42
   1727 1114 21:38:43
   2058 1114 21:38:44
   1537 1114 21:38:45
   1611 1114 21:38:46
   1231 1114 21:38:47
   1311 1114 21:38:48
    903 1114 21:38:49
   1167 1114 21:38:50
   1538 1114 21:38:51
   1203 1114 21:38:52
   1101 1114 21:38:53
   2188 1114 21:38:54
    994 1114 21:38:55
   1545 1114 21:38:56
    171 1114 21:38:57
   2106 1114 21:38:58
    810 1114 21:38:59
    635 1114 21:39:10
   1488 1114 21:39:11
   1844 1114 21:39:12
   1454 1114 21:39:13
   1170 1114 21:39:14
      1 1114 21:39:15
      1 1114 21:39:14
   1113 1114 21:39:15
      3 1114 21:39:16
      1 1114 21:39:15
      1 1114 21:39:16
      1 1114 21:39:15
   1996 1114 21:39:16
    970 1114 21:39:17
    790 1114 21:39:18
   1421 1114 21:39:19
    349 1114 21:39:20
   1317 1114 21:39:21
   1225 1114 21:39:22
   1818 1114 21:39:23
    973 1114 21:39:24
   1384 1114 21:39:25
   1747 1114 21:39:26
   1027 1114 21:39:27
    713 1114 21:39:28
   1701 1114 21:39:29
    561 1114 21:39:30
   1212 1114 21:39:31
    982 1114 21:39:32
   1621 1114 21:39:33
   1344 1114 21:39:34
   1287 1114 21:39:35
   1881 1114 21:39:36
   1518 1114 21:39:37
   1116 1114 21:39:38
   1376 1114 21:39:39
```

**Comment by porridge** [comment]
Actually, when I remember to include sort in the pipeline, the results are not that extraordinary:

```
porridge@kielonek:~/projects/k8s/55779$ egrep '^.1114 21:(38:[45]|39:[123])' kube-apiserver.log |cut -c 2-14|sort|uniq -c
   1506 1114 21:38:40
   1816 1114 21:38:41
   1292 1114 21:38:42
   1727 1114 21:38:43
   2058 1114 21:38:44
   1537 1114 21:38:45
   1611 1114 21:38:46
   1231 1114 21:38:47
   1311 1114 21:38:48
    903 1114 21:38:49
   1167 1114 21:38:50
   1538 1114 21:38:51
   1203 1114 21:38:52
   1101 1114 21:38:53
   2188 1114 21:38:54
    994 1114 21:38:55
   1545 1114 21:38:56
    171 1114 21:38:57
   2106 1114 21:38:58
    810 1114 21:38:59
    635 1114 21:39:10
   1488 1114 21:39:11
   1844 1114 21:39:12
   1454 1114 21:39:13
   1171 1114 21:39:14
   1116 1114 21:39:15
   2000 1114 21:39:16
    970 1114 21:39:17
    790 1114 21:39:18
   1421 1114 21:39:19
    349 1114 21:39:20
   1317 1114 21:39:21
   1225 1114 21:39:22
   1818 1114 21:39:23
    973 1114 21:39:24
   1384 1114 21:39:25
   1747 1114 21:39:26
   1027 1114 21:39:27
    713 1114 21:39:28
   1701 1114 21:39:29
    561 1114 21:39:30
   1212 1114 21:39:31
    982 1114 21:39:32
   1621 1114 21:39:33
   1344 1114 21:39:34
   1287 1114 21:39:35
   1881 1114 21:39:36
   1518 1114 21:39:37
   1116 1114 21:39:38
   1376 1114 21:39:39
```

**Comment by porridge** [comment]
Finally I have a lead. The number of `processing <N> initEvents of *<resource> took <above 500ms>` [warnings](https://github.com/kubernetes/kubernetes/blob/a8fc7f691f685c59e22162d2fcb7a536f2c6a108/staging/src/k8s.io/apiserver/pkg/storage/cacher.go#L947) spiked significantly right before the event. The graph shows it's usually on the order of 1 per minute, and in the critical moment it spiked to a dozen or so per minute (5 in a 10s interval):

![image](https://user-images.githubusercontent.com/489420/33371061-cc8854d0-d4f9-11e7-811e-dbc048e2d114.png)

Most of the baseline messages are about `core.Node`, while right before the event the messages about `core.Pod` spiked:
```
porridge@kielonek:~/projects/k8s/55779$ egrep processing.*initE.*Pod kube-apiserver.log
I1114 21:01:28.042284       7 cacher.go:947] processing 56729 initEvents of *core.Pod took 559.998958ms
I1114 21:26:38.043135       7 cacher.go:947] processing 79777 initEvents of *core.Pod took 612.670947ms
I1114 21:37:57.270969       7 cacher.go:947] processing 77214 initEvents of *core.Pod took 694.920593ms
I1114 21:37:57.348239       7 cacher.go:947] processing 39387 initEvents of *core.Pod took 718.415305ms
I1114 21:37:57.301393       7 cacher.go:947] processing 58038 initEvents of *core.Pod took 676.280219ms
I1114 21:37:58.875784       7 cacher.go:947] processing 66731 initEvents of *core.Pod took 835.34815ms
I1114 21:37:58.935057       7 cacher.go:947] processing 97913 initEvents of *core.Pod took 893.02411ms
I1114 21:38:01.016956       7 cacher.go:947] processing 66258 initEvents of *core.Pod took 586.854838ms
I1114 21:38:01.050750       7 cacher.go:947] processing 62421 initEvents of *core.Pod took 559.390534ms
I1114 21:38:07.125882       7 cacher.go:947] processing 90358 initEvents of *core.Pod took 1.010789013s
I1114 21:38:47.706192       7 cacher.go:947] processing 49483 initEvents of *core.Pod took 544.275345ms
I1114 21:38:47.874287       7 cacher.go:947] processing 96120 initEvents of *core.Pod took 700.552495ms
I1114 21:38:50.368708       7 cacher.go:947] processing 83178 initEvents of *core.Pod took 536.10894ms
I1114 21:38:52.754542       7 cacher.go:947] processing 53091 initEvents of *core.Pod took 635.888074ms
I1114 21:38:58.722427       7 cacher.go:947] processing 52277 initEvents of *core.Pod took 505.631515ms
I1114 21:38:58.775671       7 cacher.go:947] processing 90978 initEvents of *core.Pod took 524.691944ms
I1114 21:39:13.861891       7 cacher.go:947] processing 20929 initEvents of *core.Pod took 521.449381ms
```

@wojtek-t who investigated this in the past might have something to say here

**Comment by porridge** [comment]
After a chat with @wojtek-t one thing we established is that we think that 60-90k Pod initEvents to process is a surprisingly large number for correctness test. This would mean that some test brought up a *lot* of pods (unexpected for correctness job), or that some amount of pods violently crashlooped, or something along these lines (but then likely another test would have failed). There were indeed a few other test failures in this run, so I'll try to dig further to see if there is some correlation.

/assign

**Comment by porridge** [comment]
I looked a bit closer at the watches:

Over the course of this test job (69 minutes: 20:55:17 - 22:04:43) each node ran similar number of watch calls, 9 on average. The distribution was as follows. I think this is expected, and as intended given the jitter in watch deadlines:

|count | watches sent by a node
|----:|----:
|      1 | 7
|    533 | 8
|   2757 | 9
|   1542 | 10
|    165 | 11
|      1 | 12

However while I was expecting the watch calls to look something like the following (i.e. nice monotonic, regular growth of the rv= value)...
```
porridge@kielonek:~/projects/k8s/55779$ grep fields=spec.nodeName=gce-scale-cluster-minion-group-r30w watches
I1114 21:00:48.740608       7 get.go:238] Starting watch for /api/v1/pods, rv=147471 labels= fields=spec.nodeName=gce-scale-cluster-minion-group-r30w timeout=6m59s
I1114 21:07:47.742110       7 get.go:238] Starting watch for /api/v1/pods, rv=422358 labels= fields=spec.nodeName=gce-scale-cluster-minion-group-r30w timeout=7m57s
I1114 21:15:44.743550       7 get.go:238] Starting watch for /api/v1/pods, rv=686321 labels= fields=spec.nodeName=gce-scale-cluster-minion-group-r30w timeout=7m46s
I1114 21:23:30.746158       7 get.go:238] Starting watch for /api/v1/pods, rv=989403 labels= fields=spec.nodeName=gce-scale-cluster-minion-group-r30w timeout=8m23s
I1114 21:31:53.748492       7 get.go:238] Starting watch for /api/v1/pods, rv=1240290 labels= fields=spec.nodeName=gce-scale-cluster-minion-group-r30w timeout=5m22s
I1114 21:37:15.750371       7 get.go:238] Starting watch for /api/v1/pods, rv=1421674 labels= fields=spec.nodeName=gce-scale-cluster-minion-group-r30w timeout=7m0s
I1114 21:44:15.752103       7 get.go:238] Starting watch for /api/v1/pods, rv=1681744 labels= fields=spec.nodeName=gce-scale-cluster-minion-group-r30w timeout=8m16s
I1114 21:52:31.753769       7 get.go:238] Starting watch for /api/v1/pods, rv=2016440 labels= fields=spec.nodeName=gce-scale-cluster-minion-group-r30w timeout=5m55s
I1114 21:58:26.755795       7 get.go:238] Starting watch for /api/v1/pods, rv=2186539 labels= fields=spec.nodeName=gce-scale-cluster-minion-group-r30w timeout=9m12s
```
... what I've found instead that *only 6 nodes out of 4999* exhibit this pattern. In case of vast majority, the rv value repeats several times, it looks like this instead:
```
I1114 21:00:43.448434       7 get.go:238] Starting watch for /api/v1/pods, rv=141604 labels= fields=spec.nodeName=gce-scale-cluster-minion-group-0b7v timeout=5m47s
I1114 21:06:30.450184       7 get.go:238] Starting watch for /api/v1/pods, rv=220191 labels= fields=spec.nodeName=gce-scale-cluster-minion-group-0b7v timeout=5m11s
I1114 21:11:41.452235       7 get.go:238] Starting watch for /api/v1/pods, rv=220191 labels= fields=spec.nodeName=gce-scale-cluster-minion-group-0b7v timeout=8m26s
I1114 21:20:07.454673       7 get.go:238] Starting watch for /api/v1/pods, rv=220191 labels= fields=spec.nodeName=gce-scale-cluster-minion-group-0b7v timeout=8m16s
I1114 21:28:23.456765       7 get.go:238] Starting watch for /api/v1/pods, rv=220191 labels= fields=spec.nodeName=gce-scale-cluster-minion-group-0b7v timeout=8m35s
I1114 21:36:58.458924       7 get.go:238] Starting watch for /api/v1/pods, rv=220191 labels= fields=spec.nodeName=gce-scale-cluster-minion-group-0b7v timeout=8m51s
I1114 21:45:49.509361       7 get.go:238] Starting watch for /api/v1/pods, rv=220191 labels= fields=spec.nodeName=gce-scale-cluster-minion-group-0b7v timeout=5m47s
I1114 21:51:36.512290       7 get.go:238] Starting watch for /api/v1/pods, rv=220191 labels= fields=spec.nodeName=gce-scale-cluster-minion-group-0b7v timeout=7m33s
I1114 21:59:09.613468       7 get.go:238] Starting watch for /api/v1/pods, rv=220191 labels= fields=spec.nodeName=gce-scale-cluster-minion-group-0b7v timeout=9m41s
```

In fact the distribution of watch calls which repeat the rv= value of the previous call looks like this:
```
      6 0
      5 3
     34 4
    244 5
    904 6
   2365 7
   1285 8
    155 9
      1 10
```

Or, in other words, the distribution of number of unique (in terms of rv value) watch calls is:
```
   4005 2
    756 3
    200 4
     31 5
      1 6
      4 9
      1 10
      1 11
```

My explanation for this is that the distribution of work between nodes is extremely uneven. 94% of nodes only see a couple of pod events after startup churn, while about 0.1% of nodes do almost all the work.

This in turn I think explains the high numbers of Pod initEvents in the warnings: since most nodes are not notified of *almost any* pod events during node lifetime, every time they restart the watch, they give the same resource version number as they've seen shortly after they were initialized.

However this still does not explain why we only see the Pod initEvents warnings around the time of the cli test failure. I *suspect* that the large numbers of initEvents are seen by `cacheWatcher.process` pretty much all the time, but usually they are processed relatively quickly. The warning is logged only when there are lots of events *and* the processing takes more than 500ms. And that elevated processing time is caused by something else (my gut feeling is CPU starvation but I haven't seen evidence of this yet).

@wojtek-t do you think the above makes sense?

**Comment by wojtek-t** [comment]
Yeah - that makes perfect sense.

The only think that I may not fully agree with it is:

> (my gut feeling is CPU starvation but I haven't seen evidence of this yet

I think that we may just have been unlucky and more watches ended at that time. And processing these caused some cpu starvation. The other posibility is that other tests running at that time caused more pressure.

BTW - it's pretty interesting why only such a small percentage of nodes is receiving the load.

**Comment by porridge** [comment]
Nope, watch start rate is pretty low and even at the time of this problem (~21:39):

![watches](https://user-images.githubusercontent.com/489420/33379981-ea9908b4-d519-11e7-899f-570a1a73680b.png)

**Comment by porridge** [comment]
@kubernetes/sig-scheduling-misc is this expected that only ~0.1% of nodes are receiving pods, if there are many nodes? See https://github.com/kubernetes/kubernetes/issues/55779#issuecomment-347870686 which shows a proxy of the distribution.

**Comment by porridge** [comment]
I looked at the other failing tests in that run, and none of them looks like it could have obviously triggered this problem.

Here are some other rare messages in apiserver log, with how many times they were logged within which 10-second intervals:
```
request.go:152] Auditing failed of  request: encoding failed: policy.Eviction is not suitable for converting to "v1"
2017-11-14 21:37:00;1

upgradeaware.go:310] Error proxying data from client to backend: read tcp 10.142.0.2:443->130.211.236.253:59396: read: connection reset by peer
2017-11-14 21:35:50;1
2017-11-14 21:41:00;1
2017-11-14 21:43:50;1

watch.go:228] unable to encode watch object: client disconnected (&streaming.encoder{writer:(*framer.lengthDelimitedFrameWriter)(0xc6273716a0), encoder:(*versioning.codec)(0xc44dca4900), buf:(*bytes.Buffer)(0xc55f7a4620)})
2017-11-14 21:38:50;10

cacher.go:676] Forcing watcher close due to unresponsiveness: *core.Node
2017-11-14 21:35:50;1
2017-11-14 21:36:30;1
2017-11-14 21:37:30;1
2017-11-14 21:38:00;3
2017-11-14 21:38:30;1
2017-11-14 21:38:50;3
2017-11-14 21:39:10;3
2017-11-14 21:40:10;1
2017-11-14 21:41:10;1
2017-11-14 21:42:10;1
2017-11-14 21:43:10;1
2017-11-14 21:43:50;1
2017-11-14 21:44:50;1

watcher.go:341] Fast watcher, slow processing. Number of buffered events: 100.Probably caused by slow decoding, user not receiving fast, or other processing logic
2017-11-14 21:37:10;120
2017-11-14 21:38:50;20

watcher.go:235] Fast watcher, slow processing. Number of buffered events: 100.Probably caused by slow dispatching events to watchers
2017-11-14 21:38:00;3
2017-11-14 21:38:50;3
```

The spike of messages from `watcher.go` seems correlated in time, and supports the theory of overload. Let me try to get some latency distribution graphs...

**Comment by porridge** [comment]
I tried to look at GCE activity logs for the master VM instance, to make sure there wasn't a VM migration at the time of this problem, but unfortunately it looks like these logs are only kept for 7 days. I created an export to bigquery with a 90 day TTL so it's easier to look them up in case this re-occurs.

**Comment by porridge** [comment]
Some other findings:

4xx responses to node PATCH requests. The 409s are conflicts with node-problem-detector, which are expected. The huge 429 spikes though (the only two occurences during the entire test) confirm the overload theory.
![node-patch-4xx](https://user-images.githubusercontent.com/489420/33435430-715acb3a-d5e2-11e7-851f-ffca28ad927d.png)

**Comment by porridge** [comment]
node PATCH response latency (log scale):
![node-patch-latency](https://user-images.githubusercontent.com/489420/33435700-206da264-d5e3-11e7-80f7-0596928e0af8.png)

**Comment by porridge** [comment]
Some more latency plots (all log scale):

![put-latency](https://user-images.githubusercontent.com/489420/33435868-a028d136-d5e3-11e7-8c45-dde7bd9061d4.png)
![patch-latency](https://user-images.githubusercontent.com/489420/33435869-a0463dd4-d5e3-11e7-871d-40982936a5b3.png)
![delete-latency](https://user-images.githubusercontent.com/489420/33435870-a063ddb2-d5e3-11e7-92f4-aeb322c1b940.png)
![post-latency](https://user-images.githubusercontent.com/489420/33435872-a08233c0-d5e3-11e7-83ac-280f4b78ca33.png)
![get-latency](https://user-images.githubusercontent.com/489420/33435886-a8c10426-d5e3-11e7-96f0-0559beb7d624.png)

**Comment by porridge** [comment]
The response rates overall don't show anything terribly worrying:
![response-counts](https://user-images.githubusercontent.com/489420/33436001-04ae8506-d5e4-11e7-9b55-1d0ab292176e.png)

**Comment by porridge** [comment]
One interesting thing about the 429s is that the spikes were extremely narrow - first was ~40ms wide and the second less than 10ms wide.

**Comment by porridge** [comment]
Another interesting observation is that shortly before the event, at 21:37:48.895264 we see 'ScalingReplicaSet' Scaled up replica set kube-dns-74dbf45884 to 385 (from 313).
Then, shortly after, at 21:38:15.104060 it's scaled down to 334.

**Comment by shyamjvs** [comment]
FWIW - kube-dns replicas are controlled by kube-dns-autoscaler (which is another deployment) - https://github.com/kubernetes/kubernetes/blob/607c3d69670d624fef86ac21dd9f64ac16dde558/cluster/addons/dns/kube-dns.yaml.base#L69-L72

**Comment by porridge** [comment]
This is the kube-dns replica size in time:

![chart](https://user-images.githubusercontent.com/489420/33482504-a4447944-d699-11e7-8724-f0e2cada255d.png)

That slope is an artifact of the graphing tool - imagine it's a stepped graph for a moment.

Anyway, this is weird, because according to [the docs I've found](https://kubernetes.io/docs/tasks/administer-cluster/dns-horizontal-autoscaling/#understanding-how-dns-horizontal-autoscaling-works) the replica size should be set based on the number of nodes/cores in the cluster. So should not be changing once all nodes have been initialized.

**Comment by porridge** [comment]
@MrHohn you seem to be the owner of this addon, can you comment on the above behavior of DNS autoscaler? Could it be that the docs are outdated, and something else could influence its decisions?

**Comment by porridge** [comment]
The strangest thing is that the replica count goes *up* first. I could understand if some test does something naughty and causes some nodes to seemingly disappear from the autoscaler's radar. But this looks like some extra nodes appeared, or their shape changed!

**Comment by shyamjvs** [comment]
I would suggest looking into the dns-autoscaler pod's logs in a running cluster (we currently have one running).

**Comment by porridge** [comment]
Don't know if it's related, but the messages `gc_controller.go:62] PodGC is force deleting Pod: kube-system:e2e-image-puller-gce-scale-cluster-minion-group-*` show a similar pattern: lots of them during cluster set up, then an almost half an hour gap, and then again some of them peppered before and after the event:

```
    514  20:56
    289  20:57
    501  20:58
    699  20:59
   1434  21:00
   1398  21:01
     57  21:02
      1  21:26
      2  21:29
      3  21:30
      1  21:37
      4  21:41
      1  21:43
      1  21:45
      1  21:46
```

**Comment by shyamjvs** [comment]
I just checked the dns-autoscaler deployment on the 5k-node cluster and found that the autoscaling is configured as follows:

```
--default-params={"linear":{"coresPerReplica":256,"nodesPerReplica":16,"preventSinglePointFailure":true}}
```

16 nodesPerReplica explains why we have that value of 313 replicas for 5k nodes. Also the initial rise in your graph is reasonable given that it was probably the time when new nodes were getting registered. The less steep increase after that is the part which intrigues me.

**Comment by porridge** [comment]
The last output from the current run is roughly what I would expect it to be:
```
I1201 08:25:26.874103       1 k8sclient.go:179] Cluster status: SchedulableNodes[4936], SchedulableCores[4936]
I1201 08:25:26.874188       1 k8sclient.go:180] Replicas are not as expected : updating replicas from 305 to 309
I1201 08:25:35.246439       1 k8sclient.go:179] Cluster status: SchedulableNodes[4975], SchedulableCores[4975]
I1201 08:25:35.246537       1 k8sclient.go:180] Replicas are not as expected : updating replicas from 309 to 311
I1201 08:25:44.387838       1 k8sclient.go:179] Cluster status: SchedulableNodes[4998], SchedulableCores[4998]
I1201 08:25:44.387929       1 k8sclient.go:180] Replicas are not as expected : updating replicas from 311 to 313
porridge@gce-scale-cluster-master ~ $ date
Fri Dec  1 13:01:20 UTC 2017
porridge@gce-scale-cluster-master ~ $ 
```

**Comment by porridge** [comment]
FWIW, the bump from 313 to 385 kube-dns pods would correspond to addition of ~1150 nodes (23% increase) all in one go. The overall increase of API latency (see graphs above) is on the order of what I would expect if that many nodes were added. Except there is no evidence of additional nodes in the api server logs. `perl -n -e 'print $1,"\n" if m{(/api/v1/nodes/[a-z0-9-]+)}' kube-apiserver.log |sort|uniq -c|wc` shows 5001 (4999 + master + heapster)

**Comment by porridge** [comment]
I looked closer at [the DNS autoscaler](https://github.com/kubernetes-incubator/cluster-proportional-autoscaler/blob/master/pkg/autoscaler/k8sclient/k8sclient.go#L142) code, and there's really just a simple loop over the results of `CoreV1().Nodes().List({Watch:false})`.

It looks as if the cause for all this instability is the fact that API server for some reason returned more nodes than there in fact were. What could cause this is a mystery to me, perhaps @kubernetes/sig-api-machinery-misc could come up with some plausible scenario in which this could happen?

**Comment by porridge** [comment]
/remove-sig cli
I think it's pretty clear by now that the CLI test was just an innocent bystander.

**Comment by shyamjvs** [comment]
From the controller-manager logs, I managed to find the scaling events for the kube-dns replicaset:

```
I1114 20:54:05.874855       7 event.go:218] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"kube-system", Name:"kube-dns", UID:"f3eec23b-c97d-11e7-8959-42010a8e0002", APIVersion:"extensions", ResourceVersion:"268", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled up replica set kube-dns-74dbf45884 to 1
I1114 20:56:18.195143       7 event.go:218] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"kube-system", Name:"kube-dns", UID:"f3eec23b-c97d-11e7-8959-42010a8e0002", APIVersion:"extensions", ResourceVersion:"15994", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled up replica set kube-dns-74dbf45884 to 42
I1114 20:56:28.121041       7 event.go:218] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"kube-system", Name:"kube-dns", UID:"f3eec23b-c97d-11e7-8959-42010a8e0002", APIVersion:"extensions", ResourceVersion:"18886", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled up replica set kube-dns-74dbf45884 to 47

... increases more and more until:

I1114 21:01:29.106040       7 event.go:218] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"kube-system", Name:"kube-dns", UID:"f3eec23b-c97d-11e7-8959-42010a8e0002", APIVersion:"extensions", ResourceVersion:"184701", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled up replica set kube-dns-74dbf45884 to 311
I1114 21:01:39.933551       7 event.go:218] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"kube-system", Name:"kube-dns", UID:"f3eec23b-c97d-11e7-8959-42010a8e0002", APIVersion:"extensions", ResourceVersion:"193188", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled up replica set kube-dns-74dbf45884 to 313

but then after ~36m, we see the next event going overboard and coming back:

I1114 21:37:48.895264       7 event.go:218] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"kube-system", Name:"kube-dns", UID:"f3eec23b-c97d-11e7-8959-42010a8e0002", APIVersion:"extensions", ResourceVersion:"1500388", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled up replica set kube-dns-74dbf45884 to 385
I1114 21:38:15.104060       7 event.go:218] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"kube-system", Name:"kube-dns", UID:"f3eec23b-c97d-11e7-8959-42010a8e0002", APIVersion:"extensions", ResourceVersion:"1515810", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set kube-dns-74dbf45884 to 334
I1114 21:38:28.729751       7 event.go:218] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"kube-system", Name:"kube-dns", UID:"f3eec23b-c97d-11e7-8959-42010a8e0002", APIVersion:"extensions", ResourceVersion:"1523976", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set kube-dns-74dbf45884 to 313
I1114 21:38:39.408219       7 event.go:218] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"kube-system", Name:"kube-dns", UID:"f3eec23b-c97d-11e7-8959-42010a8e0002", APIVersion:"extensions", ResourceVersion:"1529881", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled up replica set kube-dns-74dbf45884 to 358
I1114 21:39:03.724675       7 event.go:218] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"kube-system", Name:"kube-dns", UID:"f3eec23b-c97d-11e7-8959-42010a8e0002", APIVersion:"extensions", ResourceVersion:"1544242", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled up replica set kube-dns-74dbf45884 to 385
I1114 21:39:39.507907       7 event.go:218] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"kube-system", Name:"kube-dns", UID:"f3eec23b-c97d-11e7-8959-42010a8e0002", APIVersion:"extensions", ResourceVersion:"1564496", FieldPath:""}): type: 'Normal' reason: 'ScalingReplicaSet' Scaled down replica set kube-dns-74dbf45884 to 313
```

So clearly the RS replica count was set to a higher value around that time transiently. And the count is computed by the autoscaler using #schedulable-nodes and #schedulable-cores (https://github.com/kubernetes-incubator/cluster-proportional-autoscaler/blob/master/pkg/autoscaler/controller/linearcontroller/linear_controller.go#L108). The value based on the latter is lower, so we're choosing the one based on #schedulable-nodes.

IIUC this mostly means that the no. items returned in that [list nodes call](https://github.com/kubernetes-incubator/cluster-proportional-autoscaler/blob/master/pkg/autoscaler/k8sclient/k8sclient.go#L145) were more than the no. of nodes actually in the cluster.

@porridge I suggest also checking which e2e test(s) were running in the job during that time. Maybe some test created some node objects?

**Comment by porridge** [comment]
@shyamjvs:
1. thanks for independently looking at the log, but the graph in https://github.com/kubernetes/kubernetes/issues/55779#issuecomment-348480805 already shows exactly what sizes the RS was scaled to and when. You can see it was scaled up to 385 twice, and I think this corresponds to the two latency spikes on earlier graphs.
2. I'm not sure why you think one of #shedulable-nodes and #schedulable-cores values is lower, since the log pasted in https://github.com/kubernetes/kubernetes/issues/55779#issuecomment-348489320 shows they are equal in normal situation?

I also agree that some value observed by autoscaler must have been higher. The question is why? I suspect this could be caused by:
1. actual node objects being transiently created - I don't think this is the case, since I looked at all API calls shortly before this event and there are no POSTs or PUTs to URIs including the word "node"
2. cache corruption in the autoscaler - possible,  but unlikely given how simple it is, also the apparent CPU starvation and latency hikes that API server experienced do not seem like something that could be caused by (relatively isolated) DNS autoscaler alone
3. data corruption in API server - this is my pick

FWIW, I also ran `kubectl logs -f` on the dns-autoscaler pod in today's scale-correctness test, and after the initial increase to 313 nodes it didn't budge until the master was taken down at the end of the test. So apparently whatever happened there does not happen normally in every run (or at least stopped happening since then).

**Comment by shyamjvs** [comment]
> thanks for independently looking at the log, but the graph in #55779 (comment) already shows exactly what sizes the RS was scaled to and when.

Sorry I wasn't sure from your graph if it was 'actual no. of replicas' or 'replica count set in the RS'. If it was actual no. of replicas, that doesn't necessarily mean that the count on the RS was also set the same, as it's possible that RS-controller [creates more than required replicas](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#replicas). I was confirming that the dns-autoscaler indeed set the replicas. Another possibility is that the deployment was updated (which can lead to more than desired replicas existing during the update) - but this possibility is eliminated because we see just a single RS in those events (and update means at least 2 of them).

> I'm not sure why you think one of #shedulable-nodes and #schedulable-cores values is lower, since the log pasted in #55779 (comment) shows they are equal in normal situation?

Yes, those are equal. But the values computed using them are different (one is divided by nodesPerReplica (16) and the other by coresPerReplica (256) - so that latter is lower and hence the former is picked). Ref - https://github.com/kubernetes/kubernetes/issues/55779#issuecomment-348485937

> actual node objects being transiently created - I don't think this is the case, since I looked at all API calls shortly before this event and there are no POSTs or PUTs to URIs including the word "node"

Yup.. doesn't seem to be the case as there are no such calls even in the entire apiserver logs (https://github.com/kubernetes/kubernetes/issues/55779#issuecomment-348495694).

> cache corruption in the autoscaler - possible, but unlikely given how simple it is

FMU there is no cache involved in the autoscaler here. It seems to be making just plain list calls to the apiserver? - https://github.com/kubernetes/kubernetes/issues/55779#issuecomment-348504529

> data corruption in API server - this is my pick

This might be, though I've never come across such an issue till now. If so, that sounds like a serious bug in the apiserver.

**Comment by gmarek** [comment]
When new Node is observed NodeController emits an Event:
```
Event(v1.ObjectReference{Kind:"Node", Namespace:"", Name:"gce-scale-cluster-minion-group-4-fkxb", UID:"da8413cd-c97e-11e7-8959-42010a8e0002", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'RegisteredNode' Node gce-scale-cluster-minion-group-4-fkxb event: Registered Node gce-scale-cluster-minion-group-4-fkxb in Controller
```
and logs:
```
Controller observed a new Node: "gce-scale-cluster-minion-group-2ld0"
```
In the logs from the first entry in this issue the last "new Node" was registered ~21:05, which means no new Node objects were observed afterwards.

If dns autoscaler "sees" new things it most likely mean some issue with its cache (or informer library). As we don't observe the same issues in all controllers it's very unlikely it's in the informer.

The thing I'd check is the resync logic in the dns-controller, but I won't have time to do it this weekend, nor next week:(

**Comment by porridge** [comment]
@gmarek the resync logic in dns autoscaler is [so simple](https://github.com/kubernetes-incubator/cluster-proportional-autoscaler/blob/master/pkg/autoscaler/k8sclient/k8sclient.go#L145-L157). I cannot see where it might have gone wrong...

**Comment by wojtek-t** [comment]
Yeah - dns autoscaler seems pretty simple (BTW - we should probably change it to be watch based at some point, but that's a separate issue).

That said, referring to what @shyamjvs said about apiserver cache corruption - this is not the case. If you look carefully into code of dns autoscaler, it's listing nodes directly from etcd, not from apiserver cache (it's not setting ResourceVersion=0).

So it's definitely not corruption of any cache, because there doesn't seems to be any cache involved in the whole process.

Based on the fact that there weren't any nodes created, state in etcd had to be fine.
So either:
- there was some corruption on the network between etcd<->apiserver
- or apiserver decoded/encoded it incorrectly
- there was corruption on the network between apiserver<->dns autoscaler
- something bad happened in dns-autoscaler

TBH, neither of those seem very likely to me. Corruptions on network would most probably lead to data that we can't parse. I can't imagine apiserver adding something non-existing. And DNS-autoscaler logic seems pretty simple.

I don't have any good hypothesis now. Will think about it a bit more.

**Comment by porridge** [comment]
The list of tests overlapping with the event is as follows. I'm going to take a stab at running them continuously in hope of reproducing the overload:
```
$ ./parse_build_log.py --after '2017-11-14 21:37:40'  --before '2017-11-14 21:38:02' build-log.txt 
Test 2017-11-14 21:37:24.206000->2017-11-14 21:37:40.868000 (   58 lines) [It] should be consumable from pods in volume as non-root with defaultMode and fsGroup set [Conformance]
Test 2017-11-14 21:35:33.965000->2017-11-14 21:37:43.023000 (   65 lines) [It] should support a client that connects, sends DATA, and disconnects
Test 2017-11-14 21:37:31.295000->2017-11-14 21:37:46.733000 (   46 lines) [It] Should be able to deny pod and configmap creation
Test 2017-11-14 21:31:59.634000->2017-11-14 21:37:48.435000 (  108 lines) [It] optional updates should be reflected in volume  [Conformance]
Test 2017-11-14 21:37:29.801000->2017-11-14 21:37:48.755000 (   53 lines) [It] should be consumable from pods in env vars  [Conformance]
Test 2017-11-14 21:35:23.611000->2017-11-14 21:37:54.101000 (   61 lines) [It] should replace jobs when ReplaceConcurrent
Test 2017-11-14 21:37:36.768000->2017-11-14 21:37:54.468000 (   51 lines) [It] should support (non-root,0666,default) [Conformance]
Test 2017-11-14 21:37:01.782000->2017-11-14 21:38:04.453000 (   64 lines) [It] should support a client that connects, sends DATA, and disconnects
Test 2017-11-14 21:34:10.355000->2017-11-14 21:38:11.492000 (  137 lines) [It] for RW PD with pod delete grace period of "default (30s)"
Test 2017-11-14 21:37:29.079000->2017-11-14 21:38:15.150000 (   59 lines) [It] should verify ResourceQuota with best effort scope.
Test 2017-11-14 21:36:32.206000->2017-11-14 21:38:16.496000 (   62 lines) [It] should adopt matching orphans and release non-matching pods
Test 2017-11-14 21:37:13.409000->2017-11-14 21:38:16.971000 (   52 lines) [It] should support forwarding over websockets
Test 2017-11-14 21:36:37.744000->2017-11-14 21:38:17.048000 (   78 lines) [It] should function for node-pod communication: udp  [Conformance]
Test 2017-11-14 21:36:49.147000->2017-11-14 21:38:18.370000 (   68 lines) [It] should ingest events
Test 2017-11-14 21:37:09.554000->2017-11-14 21:38:19.114000 (   45 lines) [It] should invoke init containers on a RestartAlways pod
Test 2017-11-14 21:37:20.832000->2017-11-14 21:38:20.148000 (   44 lines) [It] with readiness probe should not be ready before initial delay and never restart  [Conformance]
Test 2017-11-14 21:37:48.442000->2017-11-14 21:38:31.707000 (   55 lines) [It] should be consumable from pods in volume [Conformance]
Test 2017-11-14 21:37:48.758000->2017-11-14 21:38:34.838000 (   62 lines) [It] RecreateDeployment should delete old pods and create new ones
Test 2017-11-14 21:37:54.107000->2017-11-14 21:38:43.615000 (   55 lines) [It] should be consumable from pods in volume with defaultMode set [Conformance]
Test 2017-11-14 21:37:43.028000->2017-11-14 21:39:02.888000 (   44 lines) [It] should adopt matching pods on creation
Test 2017-11-14 21:37:46.737000->2017-11-14 21:39:05.308000 (   54 lines) [It] should provide container's memory limit [Conformance]
Test 2017-11-14 21:37:40.870000->2017-11-14 21:39:06.831000 (   64 lines) [It] should check kube-proxy urls
Test 2017-11-14 21:37:36.432000->2017-11-14 21:39:22.774000 (   98 lines) [It] should adopt matching orphans and release non-matching pods
Test 2017-11-14 21:36:46.981000->2017-11-14 21:39:41.424000 (   70 lines) [It] should *not* be restarted with a exec "cat /tmp/health" liveness probe [Conformance]
Test 2017-11-14 21:33:40.687000->2017-11-14 21:39:41.895000 (   94 lines) [It] should not schedule jobs when suspended [Slow]
Test 2017-11-14 21:37:26.288000->2017-11-14 21:39:42.359000 (   80 lines) [It] should test that deleting the Namespace of a PVC and Pod causes the successful detach of Persistent Disk
Test 2017-11-14 21:37:35.522000->2017-11-14 21:39:58.983000 (   98 lines) [It] kube-dns-autoscaler should scale kube-dns pods in both nonfaulty and faulty scenarios
Test 2017-11-14 21:37:22.238000->2017-11-14 21:40:35.494000 (  117 lines) [It] should write files of various sizes, verify size, validate content
```

**Comment by porridge** [comment]
It seems I was able to reproduce at least part of this mystery, apparently some test changes the kube-dns-autoscaler configmap, which changes the autoscaler's idea of how many pods per core it should run:

```
I1222 08:33:13.875953       1 autoscaler.go:49] Scaling Namespace: kube-system, Target: deployment/kube-dns
I1222 08:33:33.735880       1 plugin.go:50] Set control mode to linear
I1222 08:33:33.735911       1 linear_controller.go:59] ConfigMap version change (old:  new: 53805907) - rebuilding params
I1222 08:33:33.735918       1 linear_controller.go:60] Params from apiserver:
{"nodesPerReplica": 13,"coresPerReplica": 0,"min": 0,"max": 0}
I1222 08:33:33.735969       1 linear_controller.go:79] Defaulting min replicas count to 1 for linear controller
I1222 08:33:33.737528       1 k8sclient.go:179] Cluster status: SchedulableNodes[5000], SchedulableCores[5000]
I1222 08:33:33.737543       1 k8sclient.go:180] Replicas are not as expected : updating replicas from 358 to 385
I1222 08:34:14.535119       1 linear_controller.go:59] ConfigMap version change (old: 53805907 new: 53838455) - rebuilding params
I1222 08:34:14.535350       1 linear_controller.go:60] Params from apiserver:
{"coresPerReplica":256,"nodesPerReplica":16,"preventSinglePointFailure":true}
I1222 08:34:14.535496       1 linear_controller.go:79] Defaulting min replicas count to 1 for linear controller
I1222 08:34:14.537180       1 k8sclient.go:179] Cluster status: SchedulableNodes[5000], SchedulableCores[5000]
I1222 08:34:14.537296       1 k8sclient.go:180] Replicas are not as expected : updating replicas from 385 to 313
I1222 08:41:24.581800       1 linear_controller.go:59] ConfigMap version change (old: 53838455 new: 54091839) - rebuilding params
I1222 08:41:24.582174       1 linear_controller.go:60] Params from apiserver:
{"nodesPerReplica": 13,"coresPerReplica": 0,"min": 0,"max": 0}
I1222 08:41:24.582909       1 linear_controller.go:79] Defaulting min replicas count to 1 for linear controller
I1222 08:41:24.585354       1 k8sclient.go:179] Cluster status: SchedulableNodes[5000], SchedulableCores[5000]
I1222 08:41:24.585658       1 k8sclient.go:180] Replicas are not as expected : updating replicas from 313 to 385
I1222 08:41:55.875814       1 linear_controller.go:59] ConfigMap version change (old: 54091839 new: 54094960) - rebuilding params
I1222 08:41:55.876240       1 linear_controller.go:60] Params from apiserver:
{"nodesPerReplica": 15,"coresPerReplica": 15,"min": 0,"max": 0}
I1222 08:41:55.876768       1 linear_controller.go:79] Defaulting min replicas count to 1 for linear controller
I1222 08:41:55.880676       1 k8sclient.go:179] Cluster status: SchedulableNodes[5000], SchedulableCores[5000]
I1222 08:41:55.880691       1 k8sclient.go:180] Replicas are not as expected : updating replicas from 385 to 334
I1222 08:42:14.442788       1 autoscaler_server.go:133] ConfigMap not found: configmaps "kube-dns-autoscaler" not found, will create one with default params
I1222 08:42:14.448071       1 k8sclient.go:117] Created ConfigMap kube-dns-autoscaler in namespace kube-system
I1222 08:42:14.448228       1 linear_controller.go:59] ConfigMap version change (old: 54094960 new: 54122023) - rebuilding params
I1222 08:42:14.448487       1 linear_controller.go:60] Params from apiserver: 
{"coresPerReplica":256,"nodesPerReplica":16,"preventSinglePointFailure":true}
I1222 08:42:14.448815       1 linear_controller.go:79] Defaulting min replicas count to 1 for linear controller
I1222 08:42:14.450601       1 k8sclient.go:179] Cluster status: SchedulableNodes[5000], SchedulableCores[5000]
I1222 08:42:14.450758       1 k8sclient.go:180] Replicas are not as expected : updating replicas from 334 to 313
I1222 08:42:24.510299       1 linear_controller.go:59] ConfigMap version change (old: 54122023 new: 54122350) - rebuilding params
I1222 08:42:24.510624       1 linear_controller.go:60] Params from apiserver: 
{"nodesPerReplica": 14,"coresPerReplica": 0,"min": 0,"max": 0}
I1222 08:42:24.511033       1 linear_controller.go:79] Defaulting min replicas count to 1 for linear controller
I1222 08:42:24.513085       1 k8sclient.go:179] Cluster status: SchedulableNodes[5000], SchedulableCores[5000]
I1222 08:42:24.513249       1 k8sclient.go:180] Replicas are not as expected : updating replicas from 313 to 358
I1222 08:33:13.875953       1 autoscaler.go:49] Scaling Namespace: kube-system, Target: deployment/kube-dns
I1222 08:33:33.735880       1 plugin.go:50] Set control mode to linear
I1222 08:33:33.735911       1 linear_controller.go:59] ConfigMap version change (old:  new: 53805907) - rebuilding params
I1222 08:33:33.735918       1 linear_controller.go:60] Params from apiserver:
{"nodesPerReplica": 13,"coresPerReplica": 0,"min": 0,"max": 0}
I1222 08:33:33.735969       1 linear_controller.go:79] Defaulting min replicas count to 1 for linear controller
I1222 08:33:33.737528       1 k8sclient.go:179] Cluster status: SchedulableNodes[5000], SchedulableCores[5000]
I1222 08:33:33.737543       1 k8sclient.go:180] Replicas are not as expected : updating replicas from 358 to 385
I1222 08:34:14.535119       1 linear_controller.go:59] ConfigMap version change (old: 53805907 new: 53838455) - rebuilding params
I1222 08:34:14.535350       1 linear_controller.go:60] Params from apiserver:
{"coresPerReplica":256,"nodesPerReplica":16,"preventSinglePointFailure":true}
I1222 08:34:14.535496       1 linear_controller.go:79] Defaulting min replicas count to 1 for linear controller
I1222 08:34:14.537180       1 k8sclient.go:179] Cluster status: SchedulableNodes[5000], SchedulableCores[5000]
I1222 08:34:14.537296       1 k8sclient.go:180] Replicas are not as expected : updating replicas from 385 to 313
I1222 08:41:24.581800       1 linear_controller.go:59] ConfigMap version change (old: 53838455 new: 54091839) - rebuilding params
I1222 08:41:24.582174       1 linear_controller.go:60] Params from apiserver:
{"nodesPerReplica": 13,"coresPerReplica": 0,"min": 0,"max": 0}
I1222 08:41:24.582909       1 linear_controller.go:79] Defaulting min replicas count to 1 for linear controller
I1222 08:41:24.585354       1 k8sclient.go:179] Cluster status: SchedulableNodes[5000], SchedulableCores[5000]
I1222 08:41:24.585658       1 k8sclient.go:180] Replicas are not as expected : updating replicas from 313 to 385
I1222 08:41:55.875814       1 linear_controller.go:59] ConfigMap version change (old: 54091839 new: 54094960) - rebuilding params
I1222 08:41:55.876240       1 linear_controller.go:60] Params from apiserver:
{"nodesPerReplica": 15,"coresPerReplica": 15,"min": 0,"max": 0}
I1222 08:41:55.876768       1 linear_controller.go:79] Defaulting min replicas count to 1 for linear controller
I1222 08:41:55.880676       1 k8sclient.go:179] Cluster status: SchedulableNodes[5000], SchedulableCores[5000]
I1222 08:41:55.880691       1 k8sclient.go:180] Replicas are not as expected : updating replicas from 385 to 334
I1222 08:42:14.442788       1 autoscaler_server.go:133] ConfigMap not found: configmaps "kube-dns-autoscaler" not found, will create one with default params
I1222 08:42:14.448071       1 k8sclient.go:117] Created ConfigMap kube-dns-autoscaler in namespace kube-system
I1222 08:42:14.448228       1 linear_controller.go:59] ConfigMap version change (old: 54094960 new: 54122023) - rebuilding params
I1222 08:42:14.448487       1 linear_controller.go:60] Params from apiserver: 
{"coresPerReplica":256,"nodesPerReplica":16,"preventSinglePointFailure":true}
I1222 08:42:14.448815       1 linear_controller.go:79] Defaulting min replicas count to 1 for linear controller
I1222 08:42:14.450601       1 k8sclient.go:179] Cluster status: SchedulableNodes[5000], SchedulableCores[5000]
I1222 08:42:14.450758       1 k8sclient.go:180] Replicas are not as expected : updating replicas from 334 to 313
I1222 08:42:24.510299       1 linear_controller.go:59] ConfigMap version change (old: 54122023 new: 54122350) - rebuilding params
I1222 08:42:24.510624       1 linear_controller.go:60] Params from apiserver: 
{"nodesPerReplica": 14,"coresPerReplica": 0,"min": 0,"max": 0}
I1222 08:42:24.511033       1 linear_controller.go:79] Defaulting min replicas count to 1 for linear controller
I1222 08:42:24.513085       1 k8sclient.go:179] Cluster status: SchedulableNodes[5000], SchedulableCores[5000]
```

**Comment by porridge** [comment]
*Part of* mystery is solved. The good news is that the apiserver does not go crazy, and there are no phantom nodes in the system :-)

The test `kube-dns-autoscaler should scale kube-dns pods in both nonfaulty and faulty scenarios` in `test/e2e/autoscaling/dns_autoscaling.go` mutates both the autoscaler pod, and its configmap in kube-`system`. (Another test in this file is even more disruptive, but is marked as `[Serial]` so we do not run it.)

The more interesting questions now are:
1. should tests be touching `kube-system` at all? There are 36 files in `test/e2e` which mention either `kube-system` or `NamespaceSystem`, they should probably be audited. My feeling is that we should avoid it.
2. was the apiserver overload indeed caused by these rapid changes of number of kube-dns pods (which in turn were caused by the configmap changes)? I think it was, but need to reproduce it.
3. (assuming answer to (2) is "yes") should kubernetes be able to handle such changes more gracefully. I think it should.

However for the time being I think we should disable this test for large clusters.

**Comment by porridge** [comment]
I was able to reproduce some overload by running the following on an otherwise empty e2e 5k-noe cluster:

```
KUBE_GCE_ZONE=us-east1-a GINKGO_UNTIL_IT_FAILS=true kubetest -v  --test --test_args="--ginkgo.focus=kube-dns-autoscaler.should.scale.kube-dns.pods.in.both.nonfaulty.and.faulty.scenarios"
```

I wasn't able to make it as bad as to cause TLS connection establishment to fail, as seen in https://github.com/kubernetes/kubernetes/issues/55779#issuecomment-346824074 but I got some promising leads:
  - I was running `kubectl get ns -v=6` in a loop in parallel, and while most responses were in 2-3 milliseconds, I got a long tail of up to 950 milliseconds.
  - There are also plenty of these `processing N initEvents of *core.Endpoints took Xms` messages in `kube-apiserver.log`.

This at least gives us a reliable way to reproduce these problems.

**Comment by porridge** [comment]
The above error most likely corresponds to the following `kube-apiserver.log` lines:
```
I1222 11:08:31.121972       7 wrap.go:42] GET /api/v1/nodes: (25.75507948s) 200 [[e2e.test/v1.10.0 (linux/amd64) kubernetes/a7a0711] 104.132.162.81:51405]
```

```
I1222 11:08:31.121879       7 trace.go:76] Trace[858243600]: "List /api/v1/nodes" (started: 2017-12-22 11:08:05.367161191 +0000 UTC m=+96818.055062469) (total time: 25.754692024s):
Trace[858243600]: [828.64026ms] [828.633143ms] Listing from storage done
Trace[858243600]: [25.754689015s] [24.921820241s] Writing http response done (5001 items)
```

**Comment by wojtek-t** [comment]
Allocations and GC have always been significant part of profile.

This EndpointSubset isn't something I was dealing with, but it isn't completely unexpected if there are many endpoints changes in the system. Each change of "Endpoints" object is send to each KubeProxy, so in 5k-node cluster it is send (and thus copied) 5k times.

Ideally, what we should do in this place (which I never got to but always wanted to do) is to actually:
- serialize the object once
- pass the serialized form down the stack
Instead of doing deep-copy and passing a deep copy down the stack to have it serialized 5k times.

That's a larger change, but we would benefit from it a lot.

**Comment by porridge** [comment]
@wojtek-t there isn't a huge number of changes to the `kube-dns` `Endpoints` - less than 100 pods are being created/deleted in each step.

Wouldn't serializing once upfront be hard given that we don't know all API versions expected by each `kube-proxy`? Perhaps instead we could cache serialized endpoints representations (one-element cache per {`Endpoints` object, API version} pair)?

**Comment by wojtek-t** [comment]
Why you are only mentioning "kube-dns" Endpoints? The profile doesn't imply that those are only these one (I thing this includes serializing all endpoints).

> Wouldn't serializing once upfront be hard given that we don't know all API versions expected by each kube-proxy? Perhaps instead we could cache serialized endpoints representations (one-element cache per {Endpoints object, API version} pair)?

"Serialize the object once" was my "mental shortcut". Yes, you're obviously right that there might be multiple api versions, etc. So it would need to be carefully thought through what exactly we should do.
What you mentioned is one option. Another is to send down the stack pair of (serialized object, Go objects) (where serialization is to the some version and some encoding (e.g. protobuf)). And if we need a different one, than we will serialize again. Yet another one is to ensure that when a watcher registered it also given you the encoder that it is using.
There are multiple option, neither of those is trivial and obvious.

**Comment by porridge** [comment]
On Tue, Jan 2, 2018 at 1:04 PM, Wojciech Tyczynski <notifications@github.com
> wrote:

> Why you are only mentioning "kube-dns" Endpoints? The profile doesn't
> imply that those are only these one (I thing this includes serializing all
> endpoints).
>
Because kube-dns is the only deployment that is being changed at the time,
and IIUC Endpoints only changes when the set of pods returned by the parent
Service's selector changes.
Besides, all other Endpoints resources in the cluster (5 of them) are tiny
(0 or 1 element).

**Comment by wojtek-t** [comment]
But note that every change to Endpoints object is send to all kube-proxies (i.e. 5000 times). And adding 100 pods to a service may mean 100 updates (they may be batched a bit), which would mean 500k copies & serializations.

**Comment by porridge** [comment]
With #57708 fixed, I have collected two more profiles with longer duration:
 - [profile002-dns-autoscale-test.pdf](https://github.com/kubernetes/kubernetes/files/1615186/profile002-dns-autoscale-test.pdf) covers the time during which the DNS autoscaler test was running.
 -  [profile001-idle.pdf](https://github.com/kubernetes/kubernetes/files/1615183/profile001-idle.pdf) covers the same cluster being mostly idle for the same amount of time.

In the idle one, `runtime.futex()` called by memory management shows up in ~5% of the samples.
In the busy one, it is 23%. Is seems to be mostly due to mallocs invoked from 
`io/kubernetes/pkg/apis/core(*EndpointAddress)DeepCopyInto`.

This confirms the conclusions from the [previous incomplete profile](https://github.com/kubernetes/kubernetes/issues/55779#issuecomment-354452477) mentioned earlier.

**Comment by porridge** [comment]
I also graphed kube-apiserver's CPU usage (at 200ms intervals) during the above test.

![dns-test](https://user-images.githubusercontent.com/489420/34723559-5c13cb72-f54b-11e7-9a73-3ec94702a134.png)

The two significant and long usage increases correspond to the phases where `kube-dns` replica counts are increased:
 - from 313 to 385 (so by 72) and
 - from 358 to 385 (so by 27 replicas)
respectively.

The thing that puzzles me is why during the phase in between, where replica count increases from 313 to 358 - there is almost no (relatively speaking) visible CPU usage increase (even though the impact should be something in between the former two phases, judging by the fractional increase in deployment size).

**Comment by porridge** [comment]
In an attempt to understand the above I temporarily disabled the kube-dns autoscaler pod and played a little manually changing the `kube-dns` deployment size, in different increments, while measuring kube-dns CPU usage (blue line). Red line shows the replicaset size as reported by `kubectl -n kube-system get rs kube-dns-5df9d548d9 -o custom-columns=ready:status.readyReplicas --no-headers`)

In each case I first set the deployment to 300 replicas, and then increased it in the following increments, until there were 400 replicas:
 - 5

![scaleup-005](https://user-images.githubusercontent.com/489420/34723871-43f8b088-f54c-11e7-99fc-183c331fda8b.png)

 - 25

![scaleup-025](https://user-images.githubusercontent.com/489420/34723890-550fbefc-f54c-11e7-9c82-01fe48e55c45.png)


 - 50

![scaleup-050](https://user-images.githubusercontent.com/489420/34723895-5966e1f6-f54c-11e7-90c6-9910ae3f5fda.png)

 - 100 (in this last case I did it twice, until there were 500 replicas).

![scaleup-100](https://user-images.githubusercontent.com/489420/34723902-5c96893a-f54c-11e7-9ad6-631bce0d5826.png)

**Comment by porridge** [comment]
Current situation:
1. the DNS autoscaler test was disabled for large clusters today, so this should no longer cause flakes in the scale-correctness test,
2. I have filed https://github.com/kubernetes/kubernetes/issues/58050 which contains a minimal reproduction of this issue and a summary of discussions, so that people do not need to read through this huge history of debugging
With that, I'm closing this issue and fixing it will be tracked in the above one.
/close
